arangod/Pregel/Algos/AIR/AbstractAccumulator.h

```diff
@@ -0,0 +1,193 @@
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Heiko Kernbach
+/// @author Lars Maier
+/// @author Markus Pfeiffer
+///
+////////////////////////////////////////////////////////////////////////////////
+
+#ifndef ARANGODB_PREGEL_ALGOS_VERTEX_ACCUMULATORS_ABSTRACT_ACCUMULATOR_H
+#define ARANGODB_PREGEL_ALGOS_VERTEX_ACCUMULATORS_ABSTRACT_ACCUMULATOR_H 1
+#include <Pregel/Algos/AIR/Greenspun/EvalResult.h>
+#include <iostream>
+#include <numeric>
+
+#include "velocypack/Builder.h"
+#include "velocypack/velocypack-aliases.h"
+
+#include "AccumulatorOptionsDeserializer.h"
+#include "MessageData.h"
+
+namespace arangodb {
+namespace pregel {
+namespace algos {
+namespace accumulators {
+class VertexData;
+
+template <typename T>
+class Accumulator;
+
+struct AccumulatorBase {
+  AccumulatorBase(){};
+  virtual ~AccumulatorBase() = default;
+  template <typename T>
+  Accumulator<T>* castAccumulatorType() {
+    return dynamic_cast<Accumulator<T>*>(this);
+  }
+
+  enum class UpdateResult {
+    CHANGED,
+    NO_CHANGE,
+  };
+
+  // Resets the accumulator to a well-known value
+  virtual auto clear() -> greenspun::EvalResult = 0;
+  virtual auto setBySlice(VPackSlice v) -> greenspun::EvalResult = 0;
+  virtual auto getIntoBuilder(VPackBuilder& result) -> greenspun::EvalResult = 0;
+
+  // This conflates two operations: updating the accumulator, and passing the
+  // sender of the update message into the accumulator, i.e. the message passing and
+  // the accumulator operation
+  // One of these two operations should also be obsolete. This will need some consideration
+  // wrt "efficiency" (whether velocypack is the best format for messages here. It probably is,
+  // in particular since we can prevent copying stuff around)
+  virtual auto updateByMessageSlice(VPackSlice msg) -> greenspun::EvalResultT<UpdateResult> = 0;
+  virtual auto updateByMessage(MessageData const& msg) -> greenspun::EvalResultT<UpdateResult> = 0;
+
+  // used to set state on WorkerContext from message by MasterContext
+  virtual auto setStateBySlice(VPackSlice msg) -> greenspun::EvalResult = 0;
+  // used to set state on WorkerContext from message by MasterContext
+  virtual auto getStateIntoBuilder(VPackBuilder& result) -> greenspun::EvalResult = 0;
+  // used to send updates from WorkerContext to MasterContext, output of this
+  // is given to aggregateStateBySlice on MasterContext
+  virtual auto getStateUpdateIntoBuilder(VPackBuilder& result) -> greenspun::EvalResult = 0;
+  // used to aggregate states on MasterContext after receiving messages from WorkerContexts.
+  virtual auto aggregateStateBySlice(VPackSlice msg) -> greenspun::EvalResult = 0;
+
+  // Is this obsolete?
```

Comment:
I don't know if this is obsolete. But if we can answer the question now, we can either remove the function or the comment.
```suggestion
  // Is this obsolete?
```